README
COPS-Client

This module provides a simple interface to a PacketCable environment
giving a service provide an ability to set, delete and manage gates
on their CMTS equipment.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc COPS::Client

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=COPS-Client

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/COPS-Client

    CPAN Ratings
        http://cpanratings.perl.org/d/COPS-Client

    Search CPAN
        http://search.cpan.org/dist/COPS-Client/


COPYRIGHT AND LICENCE

Copyright (C) 2010 shamrock@cpan.org

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for COPS-Client

0.05	19th October 2012
	Thanks to David Wang at CableVision for finding and help
	resolve some envelope issues and TimeSet and VolumeSet problems.

	volume_set and timebase_set have been changed to be simpler as well.

	Examples are

	$cops_client->volume_set( 3000000000 );
	$cops_client->timebase_set( 60 );

	Fixed a possible non initialised variable error.
	Update 7-1 Template to include the correct pragma
	

0.04	14th May 2012
	Added support for

		6.5.6 Procedures for Modifying a Gate

		Thanks to a patch from Ivan Prostran
		
0.03    4th August 2010
        Added
	rks_set
	decode_radius_attribute

0.02	15th July 2010
	Completely made a mess of the test routines and left in tests
	which would fail on a fresh system and omitted a 64bit test

0.01    10th Jult 2010
        First version, very alpha but Man page, examples included
	See MAN page for review of bugs and current exceptions



examples/new_cops_test.pl
#!/usr/local/bin/perl

use strict;
use COPS::Client;

# ServerIP => '80.194.79.221',

my $cmts_ip = $ARGV[0];
my $action=$ARGV[1];
my $id = $ARGV[2];

my $cops_client = new COPS::Client (
                        [
                        VendorID => 'COPS Client',
                        ServerIP => $cmts_ip,
                        ServerPort => '3918',
                        Timeout => 2,
			DEBUG => 10,
			DataHandler => \&display_data,
			ListenServer => 0
                        ]
                        );

# We send a connect message to the COPS server
if ( $cops_client->connect() )
	{
	my $debug=3;

	if ( $action=~/^set$/i )
	{
	$cops_client->set_command("set");
	$cops_client->subscriber_set("ipv4","172.26.65.19");

	$cops_client->gate_specification_add(
			[
			Direction	=> 'Downstream',
			DSCPToSMark	=> 0,
			Priority	=> 0,
			PreEmption	=> 1,

			Gate_Flags	=> 0,
	                Gate_TOSField	=> 0,
			Gate_TOSMask	=> 0,
			Gate_Class	=> 0,
			Gate_T1		=> 0,
			Gate_T2		=> 0,
			Gate_T3		=> 0,
			Gate_T4		=> 0
			]
			);


	$cops_client->classifier_add(
			[
			Classifier_Type		=> 'Classifier',
			Classifier_Priority => 64,
			Classifier_SourceIP => "172.26.65.19",
			Classifier_DestinationIP => "172.26.65.1",
			Classifier_ClassifierID => 100,
			Classifier_State => 1
			]
			);

#	$cops_client->classifier_add(
#			[
#			Classifier_Type		=> 'Classifier',
#			Classifier_Priority => 64,
#			Classifier_SourceIP => "10.1.1.1",
#			Classifier_ClassifierID => 100
#			]
#			);
#
#	$cops_client->classifier_add(
#			[
#			Classifier_Type		=> 'Classifier',
#			Classifier_Priority => 64,
#			Classifier_SourceIP => "20.1.1.1",
#			Classifier_ClassifierID => 100
#			]
#			);


	$cops_client->envelope_add (
			[
			Envelope_Type		=> "authorize,reserve,commit",
			Service_Type		=> 'DOCSIS Service Class Name',
			ServiceClassName 	=> 'S_down'
			]
			);

	my $timer= time();

	$cops_client->rks_set (
			[
			PRKS_IPAddress		=> '192.168.50.2',
			PRKS_Port		=> 2000,
			PRKS_Flags		=> 1,
			SRKS_IPAddress		=> 0,
			SRKS_Port		=> 0,
			SRKS_Flags		=> 0,
			BCID_TimeStamp		=> $timer,
			BCID_ElementID		=> '99999999',
			BCID_TimeZone		=> '00000000',
			BCID_EventCounter	=> 12347890
			]
			);

#	$cops_client->opaque_set (
#			[
#			OpaqueData		=> 'a test'
#			]
#			);

#	$cops_client->timebase_set ( 
#			[
#			TimeLimit	=> 30
#			]
#			);
#

#	$cops_client->volume_set ( 
#			[
#			VolumeLimit	=> 3000
#			]
#			);


	}

	if ( $action=~/^delete$/i )
		{
	$cops_client->set_command("delete");
	$cops_client->set_gate_id($id);
	$cops_client->subscriber_set("ipv4","172.26.65.19");
		}


	if ( $action=~/^synch$/i )
	{
	$cops_client->subscriber_set("ipv4","172.26.65.19");
	$cops_client->set_command("synch");
	}

	if ( $action=~/^info$/i )
		{
		$cops_client->set_command("info");
		$cops_client->set_gate_id($id);
		$cops_client->subscriber_set("ipv4","172.26.65.19");
		}

	$cops_client->check_data_available();

	}
	else
	{
	print "Error was '".$cops_client->get_error()."'\n";
	}

$cops_client->disconnect();

sub display_data
{
my ( $self ) = shift;
my ( $data ) = shift;

print "Report Datagram sent.\n\n";

foreach my $name ( sort { $a cmp $b } keys %{$data} )
        {
        print "Name  is '$name' value is '${$data}{$name}'\n";
        }

}

exit(0);




examples/profiles/UnsolicitedGrantService.pl
#!/usr/local/bin/perl

use strict;
use COPS::Client;

my $cops_client = new COPS::Client (
                        [
                        VendorID => 'COPS Client',
                        ServerIP => '80.194.79.223',
                        ServerPort => '3918',
                        Timeout => 2,
			DataHandler => \&display_data
                        ]
                        );

if ( $cops_client->connect() )
	{
	$cops_client->set_command("set");
	$cops_client->subscriber_set("ipv4","172.26.65.19");

	$cops_client->gate_specification_add(
			[
			Direction	=> 'Upstream',
			DSCPToSMark	=> 0,
			Priority	=> 0,
			PreEmption	=> 0,

			Gate_Flags	=> 0,
	                Gate_TOSField	=> 0,
			Gate_TOSMask	=> 0,
			Gate_Class	=> 0,
			Gate_T1		=> 0,
			Gate_T2		=> 0,
			Gate_T3		=> 0,
			Gate_T4		=> 0
			]
			);


	$cops_client->classifier_add(
			[
			Classifier_Type		=> 'Classifier',
			Classifier_Priority => 64,
			Classifier_SourceIP => "172.26.65.19",
			Classifier_ClassifierID => 100,
			Classifier_State => 1
			]
			);

	$cops_client->envelope_add (
			[
			Envelope_Type		=> "authorize,reserve,commit",
			Service_Type		=> 'Unsolicited Grant Service',

			'Envelope_authorize_Request Transmission Policy' => 0,
			'Envelope_authorize_Unsolicited Grant Size' => 10000,
			'Envelope_authorize_Grants Per Interval' => 10,
			'Envelope_authorize_Nominal Grant Interval' => 1000000,
			'Envelope_authorize_Tolerated Grant Jitter' => 400000,
			'Envelope_authorize_Required Attribute Mask' => 0,
			'Envelope_authorize_Forbidden Attribute Mask' => 0,

			'Envelope_reserve_Request Transmission Policy' => 0,
			'Envelope_reserve_Unsolicited Grant Size' => 10000,
			'Envelope_reserve_Grants Per Interval' => 10,
			'Envelope_reserve_Nominal Grant Interval' => 1000000,
			'Envelope_reserve_Tolerated Grant Jitter' => 400000,
			'Envelope_reserve_Required Attribute Mask' => 0,
			'Envelope_reserve_Forbidden Attribute Mask' => 0,

			'Envelope_commit_Request Transmission Policy' => 0,
			'Envelope_commit_Unsolicited Grant Size' => 10000,
			'Envelope_commit_Grants Per Interval' => 10,
			'Envelope_commit_Nominal Grant Interval' => 1000000,
			'Envelope_commit_Tolerated Grant Jitter' => 400000,
			'Envelope_commit_Required Attribute Mask' => 0,
			'Envelope_commit_Forbidden Attribute Mask' => 0,

			]
			);

	$cops_client->check_data_available();

	}
	else
	{
	print "Error was '".$cops_client->get_error()."'\n";
	}

$cops_client->disconnect();

sub display_data
{
my ( $self ) = shift;
my ( $data ) = shift;

print "Report Information Received.\n\n";

foreach my $name ( sort { $a cmp $b } keys %{$data} )
        {
        print "Name  is '$name' value is '${$data}{$name}'\n";
        }

}

exit(0);




examples/profiles/UnsolicitedGrantServiceWithActivityDetection.pl
#!/usr/local/bin/perl

use strict;
use COPS::Client;

my $cops_client = new COPS::Client (
                        [
                        VendorID => 'COPS Client',
                        ServerIP => '80.194.79.223',
                        ServerPort => '3918',
                        Timeout => 2,
			DataHandler => \&display_data
                        ]
                        );

if ( $cops_client->connect() )
	{
	$cops_client->set_command("set");
	$cops_client->subscriber_set("ipv4","172.26.65.19");

	$cops_client->gate_specification_add(
			[
			Direction	=> 'Upstream',
			DSCPToSMark	=> 0,
			Priority	=> 0,
			PreEmption	=> 0,

			Gate_Flags	=> 0,
	                Gate_TOSField	=> 0,
			Gate_TOSMask	=> 0,
			Gate_Class	=> 0,
			Gate_T1		=> 0,
			Gate_T2		=> 0,
			Gate_T3		=> 0,
			Gate_T4		=> 0
			]
			);


	$cops_client->classifier_add(
			[
			Classifier_Type		=> 'Classifier',
			Classifier_Priority => 64,
			Classifier_SourceIP => "172.26.65.19",
			Classifier_ClassifierID => 100,
			Classifier_State => 1
			]
			);

	$cops_client->envelope_add (
			[
			Envelope_Type		=> "authorize,reserve,commit",
			Service_Type		=> 'Unsolicited Grant Service with Activity Detection',

			'Envelope_authorize_Request Transmission Policy' => 0,
			'Envelope_authorize_Unsolicited Grant Size' => 250,
			'Envelope_authorize_Grants Per Interval' => 1,
			'Envelope_authorize_Nominal Grant Interval' => 20000,
			'Envelope_authorize_Tolerated Grant Jitter' => 800,
			'Envelope_authorize_Nominal Polling Interval' => 80000,
			'Envelope_authorize_Tolerated Poll Jitter' => 40000,

			'Envelope_reserve_Request Transmission Policy' => 0,
			'Envelope_reserve_Unsolicited Grant Size' => 250,
			'Envelope_reserve_Grants Per Interval' => 1,
			'Envelope_reserve_Nominal Grant Interval' => 20000,
			'Envelope_reserve_Tolerated Grant Jitter' => 800,
			'Envelope_reserve_Nominal Polling Interval' => 80000,
			'Envelope_reserve_Tolerated Poll Jitter' => 40000,

			'Envelope_commit_Request Transmission Policy' => 0,
			'Envelope_commit_Unsolicited Grant Size' => 250,
			'Envelope_commit_Grants Per Interval' => 1,
			'Envelope_commit_Nominal Grant Interval' => 20000,
			'Envelope_commit_Tolerated Grant Jitter' => 800,
			'Envelope_commit_Nominal Polling Interval' => 80000,
			'Envelope_commit_Tolerated Poll Jitter' => 40000,

			]
			);

	$cops_client->check_data_available();

	}
	else
	{
	print "Error was '".$cops_client->get_error()."'\n";
	}

$cops_client->disconnect();

sub display_data
{
my ( $self ) = shift;
my ( $data ) = shift;

print "Report Information Received.\n\n";

foreach my $name ( sort { $a cmp $b } keys %{$data} )
        {
        print "Name  is '$name' value is '${$data}{$name}'\n";
        }

}

exit(0);




examples/profiles/NonRealTimePollingService.pl
#!/usr/local/bin/perl

use strict;
use COPS::Client;

my $cops_client = new COPS::Client (
                        [
                        VendorID => 'COPS Client',
                        ServerIP => '80.194.79.223',
                        ServerPort => '3918',
                        Timeout => 2,
			DEBUG => 5,
			DataHandler => \&display_data
                        ]
                        );

if ( $cops_client->connect() )
	{
	$cops_client->set_command("set");
	$cops_client->subscriber_set("ipv4","172.26.65.19");

	$cops_client->gate_specification_add(
			[
			Direction	=> 'Upstream',
			DSCPToSMark	=> 0,
			Priority	=> 0,
			PreEmption	=> 0,

			Gate_Flags	=> 0,
	                Gate_TOSField	=> 0,
			Gate_TOSMask	=> 0,
			Gate_Class	=> 0,
			Gate_T1		=> 0,
			Gate_T2		=> 0,
			Gate_T3		=> 0,
			Gate_T4		=> 0
			]
			);


	$cops_client->classifier_add(
			[
			Classifier_Type		=> 'Classifier',
			Classifier_Priority => 64,
			Classifier_SourceIP => "172.26.65.19",
			Classifier_ClassifierID => 100,
			Classifier_State => 1
			]
			);

	$cops_client->envelope_add (
			[
			Envelope_Type		=> "authorize,reserve,commit",
			Service_Type		=> 'Non-Real-Time Polling Service',

                        'Envelope_authorize_Traffic Priority' => 0,
			'Envelope_authorize_Request Transmission Policy' => 0,
			'Envelope_authorize_Maximum Sustained Traffic Rate' => 1000000,
			'Envelope_authorize_Maximum Traffic Burst' => 2000,
			'Envelope_authorize_Minimum Reserved Traffic Rate' => 1000000,
			'Envelope_authorize_Assumed Minimum Reserved Traffic Rate Packet Size' => 64,
			'Envelope_authorize_Maximum Concatenated Burst' => 2000,
			'Envelope_authorize_Nominal Polling Interval' => 80000,
			'Envelope_authorize_Required Attribute Mask' => 0,
			'Envelope_authorize_Forbidden Attribute Mask' => 0,

                        'Envelope_reserve_Traffic Priority' => 0,
			'Envelope_reserve_Request Transmission Policy' => 0,
			'Envelope_reserve_Maximum Sustained Traffic Rate' => 1000000,
			'Envelope_reserve_Maximum Traffic Burst' => 2000,
			'Envelope_reserve_Minimum Reserved Traffic Rate' => 1000000,
			'Envelope_reserve_Assumed Minimum Reserved Traffic Rate Packet Size' => 64,
			'Envelope_reserve_Maximum Concatenated Burst' => 2000,
			'Envelope_reserve_Nominal Polling Interval' => 80000,
			'Envelope_reserve_Required Attribute Mask' => 0,
			'Envelope_reserve_Forbidden Attribute Mask' => 0,

                        'Envelope_commit_Traffic Priority' => 0,
			'Envelope_commit_Request Transmission Policy' => 0,
			'Envelope_commit_Maximum Sustained Traffic Rate' => 1000000,
			'Envelope_commit_Maximum Traffic Burst' => 2000,
			'Envelope_commit_Minimum Reserved Traffic Rate' => 1000000,
			'Envelope_commit_Assumed Minimum Reserved Traffic Rate Packet Size' => 64,
			'Envelope_commit_Maximum Concatenated Burst' => 2000,
			'Envelope_commit_Nominal Polling Interval' => 80000,
			'Envelope_commit_Required Attribute Mask' => 0,
			'Envelope_commit_Forbidden Attribute Mask' => 0,

			]
			);

	$cops_client->check_data_available();

	}
	else
	{
	print "Error was '".$cops_client->get_error()."'\n";
	}

$cops_client->disconnect();

sub display_data
{
my ( $self ) = shift;
my ( $data ) = shift;

print "Report Information Received.\n\n";

foreach my $name ( sort { $a cmp $b } keys %{$data} )
        {
        print "Name  is '$name' value is '${$data}{$name}'\n";
        }

}

exit(0);


